### ISSUE TITLE
test_site modifies sys.path

### ISSUE THREAD

**Issue by vstinner** [python/cpython#74294]
BPO | [30108](https://bugs.python.org/issue30108)
--- | :---
Nosy | @vstinner, @serhiy-storchaka
PRs | <li>python/cpython#1197</li><li>python/cpython#1378</li><li>python/cpython#1379</li><li>python/cpython#1459</li><li>python/cpython#1460</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2017-04-20.11:50:47.947>
created_at = <Date 2017-04-20.00:23:12.110>
labels = ['3.7', 'tests']
title = 'test_site modifies sys.path'
updated_at = <Date 2017-05-04.16:52:28.706>
user = 'https://github.com/vstinner'
```

bugs.python.org fields:
```python
activity = <Date 2017-05-04.16:52:28.706>
actor = 'vstinner'
assignee = 'none'
closed = True
closed_date = <Date 2017-04-20.11:50:47.947>
closer = 'vstinner'
components = ['Tests']
creation = <Date 2017-04-20.00:23:12.110>
creator = 'vstinner'
dependencies = []
files = []
hgrepos = []
issue_num = 30108
keywords = []
message_count = 10.0
messages = ['291923', '291924', '291946', '291963', '291965', '291966', '292733', '292734', '292999', '293000']
nosy_count = 2.0
nosy_names = ['vstinner', 'serhiy.storchaka']
pr_nums = ['1197', '1378', '1379', '1459', '1460']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = None
url = 'https://bugs.python.org/issue30108'
versions = ['Python 3.7']
```

</p></details>

**Comment by vstinner** [comment]
See on Travis CI: https://travis-ci.org/python/cpython/jobs/223771666

Warning -- sys.path was modified by test_site

  Before: (47345855849656, ['', '/usr/local/lib/python37.zip', '/home/travis/build/python/cpython/Lib', '/home/travis/build/python/cpython/build/lib.linux-x86_64-3.7-pydebug'], ['', '/usr/local/lib/python37.zip', '/home/travis/build/python/cpython/Lib', '/home/travis/build/python/cpython/build/lib.linux-x86_64-3.7-pydebug'])

  After:  (47345855849656, ['', '/usr/local/lib/python37.zip', '/home/travis/build/python/cpython/Lib', '/home/travis/build/python/cpython/build/lib.linux-x86_64-3.7-pydebug'], ['', '/usr/local/lib/python37.zip', '/home/travis/build/python/cpython/Lib', '/home/travis/build/python/cpython/build/lib.linux-x86_64-3.7-pydebug', '/home/travis/.local/lib/python3.7/site-packages'])

**Comment by vstinner** [comment]
Ah, I reproduced the bug:

```sh
$ rmdir ~/.local/lib/python3.7/site-packages; ./python -m test test_site 
Run tests sequentially
0:00:00 [1/1] test_site
Warning -- sys.path was modified by test_site
  Before: (140302199734024, ['', '/home/haypo/prog/GIT/perf', '/usr/local/lib/python37.zip', '/home/haypo/prog/python/master/Lib', '/home/haypo/prog/python/master/build/lib.linux-x86_64-3.7'], ['', '/home/haypo/prog/GIT/perf', '/usr/local/lib/python37.zip', '/home/haypo/prog/python/master/Lib', '/home/haypo/prog/python/master/build/lib.linux-x86_64-3.7'])
  After:  (140302199734024, ['', '/home/haypo/prog/GIT/perf', '/usr/local/lib/python37.zip', '/home/haypo/prog/python/master/Lib', '/home/haypo/prog/python/master/build/lib.linux-x86_64-3.7'], ['', '/home/haypo/prog/GIT/perf', '/usr/local/lib/python37.zip', '/home/haypo/prog/python/master/Lib', '/home/haypo/prog/python/master/build/lib.linux-x86_64-3.7', '/home/haypo/.local/lib/python3.7/site-packages']) 
test_site failed (env changed)
```


1 test altered the execution environment:
    test_site

Total duration: 229 ms
Tests result: SUCCESS

**Comment by serhiy-storchaka** [comment]
I'm wondering if the user site directory should be removed in tearDownModule().

**Comment by vstinner** [comment]
New changeset b85c136903c6d2368162f7c4a58f258c9c69ead0 by Victor Stinner in branch 'master':
bpo-30108: Restore sys.path in test_site (bpo-1197)
https://github.com/python/cpython/commit/b85c136903c6d2368162f7c4a58f258c9c69ead0

**Comment by vstinner** [comment]
Serhiy Storchaka added the comment:
> I'm wondering if the user site directory should be removed in tearDownModule().

Writing into $HOME/.local is just completely wrong :-) The test suite
must not modify the "system".

On a buildbot, $HOME is the root directory / and so the test is
skipped since the buildbot user is not allowed to create /.local.

If you want to enhance the code: setUpModule() should mock $HOME to
use a temporary directory.

**Comment by vstinner** [comment]
The initial issue (test_site modifies sys.path) has been fixed. I don't think that it's worth it to backport the change, it's just a warning in tests.

@Serhiy: Feel free to open a new issue if you want to enhance test_site ;-)

**Comment by vstinner** [comment]
New changeset 33a5d40de997fff5b60157c546a3bf711e083bcc by Victor Stinner in branch '3.6':
bpo-30108: Restore sys.path in test_site (bpo-1197) (bpo-1378)
https://github.com/python/cpython/commit/33a5d40de997fff5b60157c546a3bf711e083bcc

**Comment by vstinner** [comment]
New changeset 092f4602c329e23b863692249cc630a3eba5b6b5 by Victor Stinner in branch '3.5':
bpo-30108: Restore sys.path in test_site (bpo-1197) (bpo-1379)
https://github.com/python/cpython/commit/092f4602c329e23b863692249cc630a3eba5b6b5

**Comment by vstinner** [comment]
New changeset 78064387e5815633168a80dcdc1bd9aec4eff46a by Victor Stinner in branch '2.7':
bpo-30108: Restore sys.path in test_site (bpo-1197) (bpo-1459)
https://github.com/python/cpython/commit/78064387e5815633168a80dcdc1bd9aec4eff46a

**Comment by vstinner** [comment]
New changeset e81e355a8e43956802211115e3f99859a1a29ecb by Victor Stinner in branch '2.7':
bpo-30108: Fix test_site setUpModule() (bpo-1460)
https://github.com/python/cpython/commit/e81e355a8e43956802211115e3f99859a1a29ecb
